### DISCUSSION TITLE
2차 스프린트는 어떻게 협업해야할까?

### DISCUSSION THREAD

**Discussion by tonic523** [woowacourse-teams/2022-smody#32] [첨예한 토론]
2차 스프린트에서 프론트와 백엔드가 페어 프로그래밍을 해야 한다는 요구사항이 주어졌습니다.
이 요구사항과 관련해 2차 스프린트에서는 우리가 어떻게 협업해야 할 지 고민이 생겼습니다.

```text
레벨3에서 가장 중요한 목적은 '찐한 협업'입니다.
그리고 이 협업은 프론트와 백엔드가 API 협업만 하는게 아니라 서로의 경계를 넘나 들며 함께 협업하는 것을 의미해요. 기획도 일부의 크루만 하는 것이 아닌 모두가 고민했으면 해서 UX워크샵도 경계 없이 다같이 참여하는 형태로 진행을 했었는데요. 그래서 실제 개발을 하는 과정에서도 FE에서 BE까지 관통하는 기능을 함께 페어 프로그래밍으로 구현하고 공유하는 경험을 만들어보면 좋을 것 같아요.
그런데 이게 막연하게 느껴지는 부분이 있어서 예시를 들자면 이런게 가능할 것 같아요.
BE의 ATDD 테스트 코드를 FE크루와 함께 개발한다.
FE의 E2E 테스트 코드를 BE크루와 함께 개발한다.
면담 신청 수정 기능을 처음부터 끝까지 FE와 BE가 함께 페어프로그래밍으로 구현한다.
페어 프로그래밍이므로 누구 한 명이 계속 드라이버를 하기 보다는, 역할을 교대한다.
BE 크루가 FE코드를 작성하기 막연하므로 FE크루가 무슨 코드를 쳐야할지 불러주고 BE크루가 타이핑을 하다보면 반복적인 패턴을 발견하면서 빠르게 학습할 수 있다.
이 외에도 서로 이야기하다보면 함께 페어 프로그래밍으로 충분히 개발할 수 있는 포인트들이 많이 있을꺼에요.
한번 이야기를 나눠보며 팀원들과 섞일 수 있는 경험을 만들어주세요.
```

> 공통

- 프론트와 백엔드가 같이 페어 프로그래밍을 하는 것이 효율적이지 않을 것 같다.

> 프론트

- 컴포넌트를 공유하는 상황에서 분업을 하게 되면 생산성이 많이 떨어질 것 같다.

> 백엔드

- 분업을 해서 서로간의 코드 리뷰를 하고 conflict를 해결하는 과정을 거쳐보고 싶다.

**Comment by tonic523** [comment]
## 방법
1. API 설계를 한다.
2. 설계에 맞게 구현한다.
  1. 백엔드는 2팀으로 나뉘어 각 팀 별로 API를 구현한다.
  2. 프론트는 1팀으로 같이 하나씩 구현한다.
3. API 설계가 바뀌어야 할 경우 다같이 모여 논의한다.

한 가지 예외 사항으로 oAuth나 배포관련 이슈는 프론트와 백엔드가 함께 페어 프로그래밍으로 진행해본다.
위 방법을 줄여서 말하자면 기존의 방식처럼 API 설계를 기준으로 프론트와 백이 나뉘어 구현하고 맞춰보는 대신 각 분야별로 분업할 수 있는 환경이라면 분업해도 되는 것 입니다!

## 이유
> 다른 방법의 시도

앞서 말한 방법이 생산성이 좋거나 효율적이라고 생각이 들지는 않아요. 기존의 방법대로 다같이 하는 것이 오히려 의견 조율에도 좋고 conflict가 발생해 git이 꼬이지 않는 확실한 방법이라고 생각해요. 그래도 시도의 관점에서는 해볼만 하다고 생각해요. 프론트와 같이 이슈까지도 분업을 하는 것은 아니지만, 백엔드끼리 분업하면서 서로간의 코드를 리뷰하고 conflict도 해결하는 경험을 할 수 있습니다.

> 프론트와 백엔드 페어 프로그래밍?

이 부분은 솔직히 효율적이지 않다고 생각이 듭니다. 그래도 이번 기회에 한 번쯤은 해볼법하지 않나? 라는 생각이 있습니다.
예외적으로 oAuth 기능은 페어로 하는게 좋을 것 같아요!

**Comment by jojogreen91** [comment]
1. 한 단위의 작업 이슈를 정한다.
2. 관련해서 FE 와 BE 가 정해야할 부분(API, 사용자 시나리오 등등) 을 협의하여 확정한다.
3. 작업한다.
 - 작업 시 백엔드는 분업하여 작업 한다. (그래서 한번의 작업 단위를 좀 크게 설정 해도 될 듯)

---

FE 의 인원을 분리하기 힘든 상황이 있기 때문에 이슈 단위로 인원을 분배하기 어렵다.
그래서 그나마 위의 방식이 동시 작업이 가능한 형태라고 생각함

**Comment by ldk980130** [comment]
## 방법
- 프론트가 같은 도메인 안에서는 분업하기 어려운 환경이기 때문에 분업을 한다고 하면 '기능 개발'/'인프라 구축'/'보안'의 세 파트로 나뉠 수 있을 것 같음
- 기능 개발할 때 논의해야 하는 서비스 시나리오 및 API 설계는 이전처럼 같이 하면 좋을 것 같음
- 프론트와 백엔드가 함께하는 페어 프로그래밍을 기능 개발의 한 파트나 보안 부분에서 하면 좋을 듯

## 이유
- 분업의 단위를 위처럼 정한 이유는 우리 프로젝트 특성상 규모가 크지 않기 때문에 '기능 개발' 안에서 분업을 할 경우 코드의 재사용성이 떨어질 수 있고 conflict이 빈번하게 날 가능성이 있기 때문이다. 분업을 한다면 아예 다른 관심사에서 하는 것이 좋을 것 같다.
- 서비스 시나리오 및 API는 결국 프로젝트의 핵심 흐름이기 때문에 모두가 함께 참여하는 것이 좋다고 생각한다.
- 프/백의 페어 프로그래밍은 해보는 것도 좋다고 생각해서 한다면 세 분업 단위 중 '보안' 쪽이 규모도 크지 않고 적절하다고 생각한다.

**Comment by ronci** [comment]
## 방법
프론트는 컨벤션, 인증 디자인(figma), 공통 컴포턴트 상의 완료 후,
백엔드는 리팩토링 및 컨벤션 상의 완료 후,
이슈별로 분업하여 코치님이 말씀하신 대로 페어프로그래밍을 진행해본다.

## 장점
1. 백엔드의 코드가 어떻게 돌아가는 지 알아보면 협업을 할 때도 더 도움이 될 수 있다.
2. 실제 협업에서 일어나는 잦은 conflict를 겪고 해결하는 것도 좋은 경험이다.
3. 코드 리뷰하면서 코드에 대해 좀 더 생각해보는 시간을 가질 수 있다.

## 단점
1. 서로의 코드를 이해하고 치는 과정이 시간 소요가 많이 된다.
2. 코드리뷰를 하면 시간이 그만큼 많이 든다.

## 결론
단점으로 시간이 많이 소요된다는 부분이 있지만
현재 과정에서는 다양한 경험을 해볼 수 있고 협업에 도움이 될 수 있도록
프론트와 백엔드가 함께 페어프로그래밍을 해보는 것이 좋다고 생각한다.
다만 프론트와 백엔드 간의 사전 상의가 모두 완료된 후 진행하는 것이 좋을 것 같다.

**Comment by woose28** [comment]
## 방법
1. API 설계를 모두 같이 한다.
2. 각 기능 별로 프론트 + 백이 짝을 지어서 구현을 한다.
  - 예를 들어 재도전 관련 기능을 빅터(프론트) + 토닉(백엔드) + 알파(백엔드)가 함께 구현한다.
3. 구현 도중 API 스펙에 변화가 있으면 다음 회의 때 변경 사항에 대해 공유를 한다.

## 이유
> 새로운 방법 시도

아직 프로젝트 초기 단계라고 생각하며, 이 시기는 다양한 방법을 시도하면서 우리 팀에게 어울리는 협업 방법을 찾는 시기라 생각한다.
따라서 새로운 협업 방식을 도입해서 우리 팀에게 분업 방식이 어울리는지 평가해 보면 좋겠다.

## 전제 조건
위에서 제안한 분업 방식을 시행하기 위해서 프론트에서 UI 시안과 컨벤션, 공통 컴포넌트 개발 시의 규칙 등의 협의돼야 한다.

**Comment by bcc0830** [comment]
<h2>방법</h2>
1. api 명세는 다같이 진행한다 <br>
2. 작업은 팀 단위가 아니라 이슈 단위로 진행한다<br>
3. 프론트의 경우는 굳이 분리하지 않고, 백은 2인 1조로 분리한다<br>
4. 기능을 구현한다<br>

<h2>장점</h2>
1. 프론트의 경우는 공통적인 부분을 같이 개발할 수 있어서 재활용에 유리<br>
2. 백엔드의 경우 팀이 나누어지기 때문에 백엔드만의 로직은 빨리 개발할 수 있다<br>
3. 백엔드의 경우 어느정도 컨벤션이 있기 때문에 공유만 잘 된다면 생각보다 컨벤션 이슈에 드는 시간이 적을수도 있다(뇌피셜)<br>
4. 일단 각자의 진영의 입장을 존중하기 때문에 시작이 빠르다 <br>

<h2>단점</h2>
1. 백엔드의 경우는 프론트의 작업속도보다 빠를 수 있기 때문에 백엔드 2팀 중 한 팀은 실질적으로 연결해서 테스트해볼 수 없다<br>
2. 백엔드의 경우는 이미 구현이 완료된 부분에 프론트가 구현을 시작할 시 변동사항이 발생할 수 있고, 이미 완료된 기능을 수정해야 하므로 자칫 일을 2번 하게 되는 느낌을 받을 수도 있다<br>

**Comment by wonsss** [comment]
## 방법
1. 이번 스프린트 목표를 수립한다.
2. 목표를 이루기 위한 이슈를 만든다.
3. 전체 팀 회의를 통해 이슈를 담당자에게 할당한다.(이슈 규모에 따라 유연하게 인원 배정)
4. 이슈 데드라인을 정한다.
5. 이슈 담당자들이 해당 이슈에 관련된 API와 디자인을 기획하고, 전체 팀원에게 공유 후 피드백 수렴을 당일에 마감한다.
6. 이슈 담당자들끼리 구현을 진행한다.

- 페어프로그래밍 방식만을 굳이 고수하지 않고 분업 등을 상황에 맞춰 유연하게 적용한다.
- 프론트는 분업 전에 공통 디자인 시스템을 만들어야 한다.

- 장점 : 효율적인 개발
- 단점 : 컨플릭트 우려
